### ISSUE TITLE
Object creation benchmark

### ISSUE THREAD

**Issue by mcollina** [davidmarkclements/v8-perf#8]
@hashseed thanks so much for in-depth review and fixes:

in https://github.com/nodejs/CTC/issues/155#issuecomment-318670712 you wrote:

> Object creation benchmarks don't actually create the objects because TF optimized it away. Again, with [my changes](https://github.com/TimothyGu/v8-perf/commit/f79a270e89d8a5eb296f338ac5f950e40d7ec251#diff-1b6ee64f4585ff309447c26b9c6d4d11), the benchmark results are more useful:

   Node.js v8.2.1 with V8 5.8:

   ```
   literal x 28,455,810 ops/sec ±1.19% (83 runs sampled)
   class x 21,993,317 ops/sec ±1.13% (86 runs sampled)
   constructor x 28,093,899 ops/sec ±0.73% (82 runs sampled)
   ```

   Node.js v9.0.0-v8-canary20170724c045f1f8dc with V8 6.2.4:

   ```
   literal x 31,861,977 ops/sec ±0.88% (86 runs sampled)
   class x 32,265,246 ops/sec ±0.65% (90 runs sampled)
   constructor x 32,120,661 ops/sec ±0.99% (87 runs sampled)
   ```

When is this optimization happening? I've tried https://gist.github.com/mcollina/b903a1ef99f872ceb8e51dc491be34b8 (accessing a property) and https://gist.github.com/mcollina/40e91ad55ae9c3b0131c2a8b79387dfe (passing an object to a function), and it leads to similar previous results compared to the previous benchmark. However https://gist.github.com/mcollina/0bbcca999f7c190d97009165b7913824 reduces the gap.

It seems that if the object has hardcoded properties, then Turbofan can optimize the object creation away. I think this is a change worth including in our report, as passing objects with hardcoded properties as an "option object" is a widely used pattern.

I would like to keep including the previous benchmark (or one of above) and rebrand it as "option objects", to document this new behavior which in my opinion it is a fantastic result.
Which benchmark do you think fit better the case?

cc @bmeurer @TimothyGu

**Comment by TimothyGu** [comment]
To be clear, the fact that the object creation benchmark doesn't test object creation was actually discovered by Jakob Kummerow of the V8 project in a comment to the blog post, and it was I who wrote the fix and https://github.com/nodejs/CTC/issues/155#issuecomment-318670712 you linked above. I'm not sure if he is on GitHub though.

Speaking of that, it seems like there was some merge conflicts: https://github.com/davidmarkclements/v8-perf/blob/master/bench/object-creation.js now again contains a `var obj`…

**Comment by mcollina** [comment]
> Speaking of that, it seems like there was some merge conflicts: https://github.com/davidmarkclements/v8-perf/blob/master/bench/object-creation.js now again contains a var obj…

It seems `Object.create()` is not optimized away. I will remove the `var` anyway for parity.

**Comment by jakobkummerow** [comment]
+1 to @TimothyGu 's changes. The new results look plausible.

> Turbofan can optimize the object creation away

Crankshaft could do that too, under the right circumstances ;-)

**Comment by davidmarkclements** [comment]
the `Object.create` bench isn't equivalent - it's creating a object with a prototype of `{ x: 1 }`

either we have to set this with define properties config object or by adding a prop

**Comment by davidmarkclements** [comment]
hey @jakobkummerow - which circumstances can trigger this object inlining behaviour in Crankshaft?

**Comment by jakobkummerow** [comment]
> which circumstances can trigger this object inlining behaviour in Crankshaft?

Probably mostly the circumstances in the Octane benchmark ;-)
Frankly I'm surprised that the first version of your benchmark didn't seem to trigger the optimization in Crankshaft. I don't have time today to compile an old version and play with it though.

**Comment by mcollina** [comment]
Fixed in #17.
